app/assets/javascripts/tylium/modules/auto_save/local.js

```diff
@@ -0,0 +1,144 @@
+        localStorage.removeItem(that.key);
+      })
+    }
+
+    // Find all inputs in the form, then exclude based on excluded input types
+    var formInputs = Array.from(this.target.querySelectorAll('input')).filter(function(input) {
+      return that.permittedInputTypes.includes(input.getAttribute('type')) && !that.excludedInputNames.includes(input.name);
+    })
+
+    formInputs = formInputs.concat(Array.from(this.target.querySelectorAll('textarea, select')));
+
+    formInputs.forEach(function(input) {
+      // we're using a jQuery plugin for :textchange event, so need to use $()
+      $(input).on('textchange change', that.debounce(that.setData.bind(that), 500));
+    })
+  }
+
+  debounce(func, delay) {
+    let debounceTimer;
+
+    return function () {
+      const context = this;
+      const args = arguments;
+      clearTimeout(debounceTimer);
+      debounceTimer = setTimeout(() => func.apply(context, args), delay);
+    };
+  }
+
+  getData() {
+    var that = this;
+    var data = new FormData(this.target);
+    var dataArray = Array.from(data).filter(function(dataElement) {
+      return !that.excludedInputNames.includes(dataElement[0] );
+    });
+
+    return dataArray;
+  }
+
+  restoreData() {
+    var that = this;
+    var data = JSON.parse(localStorage.getItem(this.key));
+
+    if (data !== null) {
+      // Deserialization of form data
+      // Reference: https://wildwolf.name/store-formdata-object-in-localstorage/
+      for (var item of data) {
+        var name = item[0];
+        var element   = that.target.elements[name];
+
+        // Skip loop if element is undefined
+        if (!element) { continue; }
+
+        // RadioNodeList is an array of checkboxes or radio buttons
+        if (element instanceof RadioNodeList) {
+          element.forEach(function(radioNodeListElement) {
+            if (['checkbox', 'radio'].includes(radioNodeListElement.type)) {
+              if (radioNodeListElement.value === item[1]) {
+                radioNodeListElement.checked = true;
+              }
+            }
+          })
+        } else if (element.type === 'checkbox') {
+          if (element.value === item[1]) {
+            element.checked = true;
+          }
+        } else if (element.type === 'file') {
+          element.value = '';
+        } else {
+          element.value = item[1];
+          $(element).trigger('load-preview');
+        }
+
+        this.restoreTagInputDisplay(data);
+      }
+    }
+  }
+
+  restoreTagInputDisplay(data) {
+    var tagInputData = data.find(function(item) {
+      return item[0] === 'issue[tag_list]';
+    })
+
+    if (tagInputData) {
+      var hiddenInput = document.querySelector('#issue_tag_list');
+      hiddenInput.value = tagInputData[1];
+
+      var dropdownBtnSpan = document.querySelector('#issues_editor .dropdown-toggle span.tag');
+      var selectedDropdown = document.querySelector(`.js-taglink[data-tag='${tagInputData[1]}']`);
+
```

Comment:
I don't love that we still duplicate this functionality. We can extract it and re-use it.

```coffeescript
window.tagLink = (target)->
  $target = $(target)
  $('#issue_tag_list').val($target.data('tag'))
  $span = $("#issues_editor .dropdown-toggle span.tag")
  $span.html($target.html())
  $span.css("color", $target.css("color"))
  $('#issue_tag_list').trigger('textchange')

document.addEventListener "turbolinks:load", ->
  $('#issues_editor .js-taglink').click (e) ->
    tagLink(e.target)
```

```javascript
    if (tagInputData) {
      var selectedDropdown = document.querySelector(`.js-taglink[data-tag='${tagInputData[1]}']`);
      window.tagLink(selectedDropdown);
    }
```

That's an idea. It's a little ugly for us to reach out and use the global function. So I'd hear an argument against it.

Reply:
What about moving it into a class? I've tested and it works as well.
```
class @SelectTagDropdown
  constructor: (@$target) ->
    @init()

  init: ->
    $span = $("#issues_editor .dropdown-toggle span.tag")
    $span.css("color", @$target.css("color"))
    $span.html(@$target.html())

document.addEventListener "turbolinks:load", ->
  $('#issues_editor .js-taglink').click (e) ->
    $target = $(e.target)
    new SelectTagDropdown($target)
    $('#issue_tag_list').val($target.data('tag'))
    $('#issue_tag_list').trigger('textchange')
```

in `auto_save/local.js`
```
    if (tagInputData) {
      var hiddenInput = document.querySelector('#issue_tag_list');
      hiddenInput.value = tagInputData[1];

      var $target = $(`.js-taglink[data-tag='${tagInputData[1]}']`);
      new SelectTagDropdown($target)
    }
```